sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/manager/DownloadManagerReleaseDownloader.java

```diff
@@ -0,0 +1,253 @@
+            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN);
+            request.setVisibleInDownloadsUi(false);
+        }
+        long enqueueTime = System.currentTimeMillis();
+        long downloadId = downloadManager.enqueue(request);
+
+        /* Check for if state changed and task not canceled in time. */
+        if (mRequestTask == task) {
+
+            /* Delete previous download. */
+            long previousDownloadId = SharedPreferencesManager.getLong(PREFERENCE_KEY_DOWNLOAD_ID, INVALID_DOWNLOAD_IDENTIFIER);
+            if (previousDownloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+                AppCenterLog.debug(LOG_TAG, "Delete previous download id=" + previousDownloadId);
+                downloadManager.remove(previousDownloadId);
+            }
+
+            /* Store new download identifier. */
+            setDownloadId(downloadId);
+            mListener.onStart(enqueueTime);
+
+            /* Start monitoring progress for mandatory update. */
+            if (mReleaseDetails.isMandatoryUpdate()) {
+                update();
+            }
+        } else {
+
+            /* State changed quickly, cancel download. */
+            AppCenterLog.debug(LOG_TAG, "State changed while downloading, cancel id=" + downloadId);
+            downloadManager.remove(downloadId);
+        }
+    }
+
+    @WorkerThread
+    synchronized void onUpdate() {
+
+        /* Query download manager. */
+        DownloadManager downloadManager = getDownloadManager();
+        try {
+            Cursor cursor = downloadManager.query(new DownloadManager.Query().setFilterById(mDownloadId));
+            if (cursor == null) {
+                throw new NoSuchElementException();
+            }
+            try {
+                if (!cursor.moveToFirst()) {
+                    throw new NoSuchElementException();
+                }
+                int status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS));
+                if (status == DownloadManager.STATUS_FAILED) {
+                    throw new IllegalStateException();
+                }
+                if (status != DownloadManager.STATUS_SUCCESSFUL) {
+                    long totalSize = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));
+                    long currentSize = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
+                    if (mListener.onProgress(currentSize, totalSize)) {
+
+                        /* Schedule the next check if more updates are needed. */
+                        HandlerUtils.getMainHandler().postAtTime(new Runnable() {
+
+                            @Override
+                            public void run() {
+                                update();
+                            }
+                        }, HANDLER_TOKEN_CHECK_PROGRESS, SystemClock.uptimeMillis() + CHECK_PROGRESS_TIME_INTERVAL_IN_MILLIS);
```

Comment:
CHECK_PROGRESS_TIME_INTERVAL_IN_MILLIS should be merged into the android 4 setting and decide a common value for both. Both control frequency of update of the progress dialogs.